### PR TITLE
doc(LifecycleHooks): update API doc

### PR DESCRIPTION
**Pull request by vicb**
first commit is #4337, second commit has the doc updates

### CODE REVIEW

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 1]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
- *     if (changes['propA']) {
- *       // if propA was updated
- *     }
- *     if (changes['propA']) {
- *       // if propB was updated
- *     }
+ * ```typescript
+ * @Component({selector: 'my-cmp'})
+ * @View({template: `<p>prop = {{prop}}</p>`})
+ * class MyComponent implements OnInit, OnDestroy, OnChanges {
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: `OnInit`, `OnDestroy` are not implemented (nor needed for the example).

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 2]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  * - `AfterContentInit`,
  * - `AfterContentChecked`,
+ * - `AfterViewInit`,
+ * - `AfterViewChecked`,
  * - `OnDestroy` (at the very end before destruction)
  */
 
 /**
- * Notify a directive when any of its bindings have changed.
+ * Implement this interface to get notified when any data bound property of this directives changes.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _this directives_   -->  _the directive_ (?)

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 3]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
 
 /**
- * Notify a directive when it has been checked the first time.
+ * Implement this interface to get notified when the directive has been checked for the first time.
  *
  * `onInit` is called right after the directive's bindings have been checked for the first time,
  * and before any of its children's bindings have been checked.
  *
- * It is invoked only once.
+ * It is invoked only once when the directive is first instantiated
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: Missing `.` at the end.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 4]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * ### Example ([live example](http://plnkr.co/edit/1MBypRryXd64v4pV03Yn?p=preview))
  *
- * ```
- * @Component(...)
- * class MyComponent implements OnInit {
- *   onInit(): void {
+ * ```typescript
+ * @Component({selector: 'my-cmp'})
+ * @View({template: `<p>prop = {{prop}}</p>`})
+ * class MyComponent implements OnInit, OnDestroy, OnChanges {
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: `OnDestroy`, `OnChanges` not implemented/needed.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 5]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *   displayChild = true;
+ * }
+
  *  ```
  */
 export interface OnInit { onInit(); }
 
 /**
- * Overrides the default change detection.
+ * Implements this interface to override the default change detection algorithm for the directive.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _Implements_  -->  _Implement_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 6]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * export class App {
+ *   list = [];
+ * }
+ * ```
  */
 export interface DoCheck { doCheck(); }
 
 /**
- * Notify a directive whenever a {@link ViewMetadata} that contains it is destroyed.
+ * Implements this interface to get notified when the directive is destroyed, that is when the
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _Implements_  -->  _Implement_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 7]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * ```
- * @Component(...)
- * class MyComponent implements OnDestroy {
- *   onDestroy(): void {
- *     // invoked to notify directive of the containing view destruction.
+ * ```typesript
+ * @Component({selector: 'my-cmp'})
+ * @View({template: `<p>prop = {{prop}}</p>`})
+ * class MyComponent implements OnInit, OnDestroy, OnChanges {
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: `OnInit`, `OnChanges` is not implemented/needed.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 8]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * }
+
  * ```
  */
 export interface OnDestroy { onDestroy(); }
 
 /**
- * Notify a directive when the bindings of all its content children have been checked the first
- * time (whether they have changed or not).
+ * Implements this interface to get notified when any data-bound property of all the directive
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _Implements_  -->  _Implement_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 9]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * content children have been checked for the first time.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _have_  -->  _has_
<sub>(Although the whole sentence is not very clear to me anyway.)</sub>

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 10]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * ## Example
+ * `afterContentInit` gets called whether content children have changed or not.
  *
- * ```
+ * ### Example
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: Where's the plnkr :stuck_out_tongue:

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 11]
```diff
@@ -29,103 +36,192 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *
+ * ```typescript
  * @Component(...)
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: Not directly related to this PR, but aren't the examples supposed to be actually runnable ?

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 12]
```diff
@@ -136,28 +232,65 @@ export interface OnDestroy { onDestroy(); }
 export interface AfterContentInit { afterContentInit(); }
 
 /**
- * Notify a directive when the bindings of all its content children have been checked (whether
- * they have changed or not).
+ * Implements this interface to get notified when any data-bound property of all the directive
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _Implements_  -->  _Implement_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 13]
```diff
@@ -136,28 +232,65 @@ export interface OnDestroy { onDestroy(); }
...
+ * content children have been checked.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _have_  -->  _has_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 14]
```diff
@@ -136,28 +232,65 @@ export interface OnDestroy { onDestroy(); }
...
+ * @Component({selector: 'child-cmp'})
+ * @View({template: `{{where}} child`})
+ * class ChildComponent {
+ *   @Property() where: string;
+ * }
+ *
+ * @Component({selector: 'parent-cmp'})
+ * @View({
+ *   template: `<ng-content></ng-content>`,
+ *     directives: [ChildComponent]
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: *indentation*

Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: Is the `directives` property needed anyway (since there is no `<child-cmp>` in the template) ?

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 15]
```diff
@@ -136,28 +232,65 @@ export interface OnDestroy { onDestroy(); }
...
+ *   directives: [NgIf, ParentComponent, ChildComponent]
+ * })
+ * *  ```
  */
 export interface AfterContentChecked { afterContentChecked(); }
 
 /**
- * Notify a directive when the bindings of all its view children have been checked the first time
- * (whether they have changed or not).
+ * Implements this interface to get notified when any data-bound property of all the directive
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _Implements_  -->  _Implement_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 16]
```diff
@@ -136,28 +232,65 @@ export interface OnDestroy { onDestroy(); }
...
+ * view children have been checked for the first time.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _have_  -->  _has_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 17]
```diff
@@ -168,17 +301,51 @@ export interface AfterContentChecked { afterContentChecked(); }
 export interface AfterViewInit { afterViewInit(); }
 
 /**
- * Notify a directive when the bindings of all its view children have been checked (whether they
- * have changed or not).
+ * Implements this interface to get notified when any data-bound property of all the directive
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: For the last time: _Implements_  -->  _Implement_ :stuck_out_tongue:

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 18]
```diff
@@ -168,17 +301,51 @@ export interface AfterContentChecked { afterContentChecked(); }
...
+ * view children have been checked.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _have_  -->  _has_

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 19]
```diff
@@ -168,17 +301,51 @@ export interface AfterContentChecked { afterContentChecked(); }
...
+ * class ChildComponent {
+ *   @Property() where: string;
+ * }
+ *
+ * @Component({selector: 'parent-cmp'})
+ * @View({
+ *   template: `
+ *     <button (click)="showView = !showView">Toggle view child</button>
+ *     <child-cmp *ng-if="showView" where="view"></child-cmp>
+ *     directives: [NgIf, ChildComponent]
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: Missing trailing `` `,`` in the template and `directives` is too indented.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 20]
```diff
@@ -168,17 +301,51 @@ export interface AfterContentChecked { afterContentChecked(); }
...
+ * })
+ * class ParentComponent implements AfterViewChecked, AfterContentChecked {
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: `AfterContentChecked` is not implemented/needed.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 21]
```diff
@@ -168,17 +301,51 @@ export interface AfterContentChecked { afterContentChecked(); }
...
+ *   private getMessage(cmp: ChildComponent): string {
+ *     return cmp ? cmp.where + ' child' : 'no child';
  *   }
  * }
+ *
+ * @Component({selector: 'app'})
+ * @View({
+ *   template: `<parent-cmp></parent-cmp>`,
+ *   directives: [ParentComponent]
+ * })
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: The actual class is missing.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 22]
```diff
@@ -136,28 +232,65 @@ export interface OnDestroy { onDestroy(); }
...
+ * @Component({selector: 'app'})
+ * @View({
+ *   template: `
+ *   <parent-cmp>
+ *     <button (click)="showContent = !showContent">Toggle content child</button>
+ *     <child-cmp *ng-if="showContent" where="content"></child-cmp>
+ *   </parent-cmp>`,
+ *   directives: [NgIf, ParentComponent, ChildComponent]
+ * })
+ * *  ```
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: The `App` class is missing (plus one `*` too many).

**Comment by PascalPrecht** [comment]
I think I just realised that @gkalpak already mentioned most of the typos that I mentioned on the PR. Didn't want to comment just on the comment. Sorry for spamming

**Comment by vicb** [comment]
Thank you guys for the feedback, PTAL

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 23]
```diff
@@ -30,156 +36,393 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  * - `AfterContentInit`,
  * - `AfterContentChecked`,
+ * - `AfterViewInit`,
+ * - `AfterViewChecked`,
  * - `OnDestroy` (at the very end before destruction)
  */
 
 /**
- * Notify a directive when any of its bindings have changed.
+ * Implement this interface to get notified when any data-bound property of this directives changes.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _this directives_  -->  _the directive_ (?)

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 24]
```diff
@@ -30,156 +36,393 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  */
 export interface OnChanges { onChanges(changes: StringMap<string, SimpleChange>); }
 
 /**
- * Notify a directive when it has been checked the first time.
+ * Implement this interface to get notified when the directive has been checked for the first time.
  *
- * `onInit` is called right after the directive's bindings have been checked for the first time,
- * and before any of its children's bindings have been checked.
+ * `onInit` is called right after the directive's has been checked for the first time, and before
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: _the directive's_  -->  _the directive's data-bound properties_ (?)

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 25]
```diff
@@ -30,156 +36,393 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * @Component({selector: 'child-cmp'})
+ * @View({template: `{{where}} child`})
+ * class ChildComponent {
+ *   @Property() where: string;
+ * }
+ *
+ * @Component({selector: 'parent-cmp'})
+ * @View({
+ *   template: `<ng-content></ng-content>`,
+ *   directives: [ChildComponent]
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: I think I've asked again, but here I am again :smiley:: Is this necessary here ?

Comment (vicb) [inline modules/angular2/src/core/compiler/interfaces.ts]: It's not. Fixed here & in plunker - as well as all your other feedback items. Thanks again!

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 26]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *   onDestroy() {
+ *     console.log('onDestroy');
  *   }
  * }
- * ```
+ *
+ * @Component({selector: 'app'})
+ * @View({
+ *   template: `
+ *     <button (click)="hasChild = !hasChild">7
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/interfaces.ts]: Hey, there's a sneaky `7` at the end ?!
<sub>(It isn't there on purpose, is it ?)</sub>

Comment (vicb) [inline modules/angular2/src/core/compiler/interfaces.ts]: The 7 key is too close to the up arrow on my keyboard...
Thanks for noticing, I'll update

**Comment by vicb** [comment]
@vsavkin or @rkirov would you have time to review this ?

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 27]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * Implement this interface to get notified when any data-bound property of the directive changes.
  *
- * `onChanges` is called right after the directive's bindings have been checked,
- * and before any of its children's bindings have been checked.
+ * `onChanges` is called right after the data-bound properties have been checked and before view
+ * and content children are checked.
  *
- * It is invoked only if at least one of the directive's bindings has changed.
+ * It is invoked after all the directive's data-bound properties have been updated if at least one
+ * of them has changed.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: this repeats the first sentence. remove

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 28]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * ```
- * @Component(...)
+ * ```typescript
+ * @Component({selector: 'my-cmp'})
+ * @View({template: `<p>prop = {{prop}}</p>`})
  * class MyComponent implements OnChanges {
- *   propA;
- *   propB;
+ *   @Property() prop: any;
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: can you call this myProp or something else that can clearly be distinguished from a keyword (the goal is to make it clear that the name is user-defined)

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 29]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * }
+ *
+ * bootstrap(App).catch(err => console.error(err));
+ * ```
  */
 export interface OnChanges { onChanges(changes: StringMap<string, SimpleChange>); }
 
 /**
- * Notify a directive when it has been checked the first time.
+ * Implement this interface to get notified when the directive has been checked for the first time.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: "checked" is a bit ambiguous without more context. maybe reword as:

```
Implement this interface to get notified when the data-bound properties of a directive/component instance have been checked by Angular for the first time since instantiation.
```

what do you think?

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 30]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * `onInit` is called right after the directive's bindings have been checked for the first time,
- * and before any of its children's bindings have been checked.
+ * `onInit` is called right after the directive's data-bound properties have been checked for the
+ * first time, and before any of its children has been checked.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: has -> have

File: modules/angular2/src/core/compiler/interfaces.ts (hunk 29 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: ok, in the light of the content below, this is the alternative version:

----
Implement this interface to get notified when the directive or component's data-bound properties have been initialized.

----

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 31]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * It is invoked only once.
+ * It is invoked only once when the directive is instantiated.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: append this to the previous paragraph, no need for a new one.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 32]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
- * and before any of its children's bindings have been checked.
+ * `onChanges` is called right after the data-bound properties have been checked and before view
+ * and content children are checked.
  *
- * It is invoked only if at least one of the directive's bindings has changed.
+ * It is invoked after all the directive's data-bound properties have been updated if at least one
+ * of them has changed.
  *
- * ## Example:
+ * ### Example ([live example](http://plnkr.co/edit/AHrB6opLqHDBPkt4KpdT?p=preview)):
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: can you also add info about the value passed into the callback?

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 33]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
 /**
- * Overrides the default change detection.
+ * Implement this interface to override the default change detection algorithm for the directive.
  *
- * `doCheck()` gets called to check the changes in the directives instead of the default
- * change detection mechanism.
+ * `doCheck` gets called to check the changes in the directives instead of the default algorithm.
  *
- * It is invoked every time the change detection is triggered.
+ * The default change detection algorithm looks for differences by comparing property values by
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: property -> bound property

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 34]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * reference across change detection runs. When `DoCheck` is implemented, the default algorithm is
+ * disabled and `doCheck` is responsible for checking for changes.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: we should also explain why would you want to do this. primarily to improve performance by using insights about the component, it's implementation and data types of its properties.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 35]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * ## Example
+ * Use {@link KeyValueDiffers} and {@link IterableDiffers} to add your custom check mechanisms.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: how does `doCheck` interact with `onChanges`? does that still get called or should the "reaction" to the detected changes be handled within the doCheck method?

also does doCheck need to signal back to angular if any changes have been found?

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 36]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * ```
- * @Component(...)
- * class MyComponent implements DoCheck {
- *   doCheck(): void {
- *     // Custom logic to detect changes
+ * ### Example ([live demo](http://plnkr.co/edit/QpnIlF0CR2i5bcYbHEUJ?p=preview))
+ *
+ * In the following example `doCheck` uses an `IterableDiffer` to detect the updates to the
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: add a link to IterableDiffers

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 37]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *   list = [];
+ * }
+ * ```
  */
 export interface DoCheck { doCheck(); }
 
 /**
- * Notify a directive whenever a {@link ViewMetadata} that contains it is destroyed.
+ * Implement this interface to get notified when the directive is destroyed, that is when the
+ * containing {@link ViewMetadata} is destroyed.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: this has nothing to do with ViewMetadata. it should read "Implement this interface to get notified when a directive/component instance is destroyed."

Also add "`onDestroy` callback is typically used for any custom cleanup that needs to occur when the instance is destroyed".

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 38]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *
+ * bootstrap(App).catch(err => console.error(err));
+ * * ```
  */
 export interface OnDestroy { onDestroy(); }
 
 /**
- * Notify a directive when the bindings of all its content children have been checked the first
- * time (whether they have changed or not).
+ * Implement this interface to get notified when all of the content children have been checked for
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: "have been checked for the first time." -> "have had their data-bound properties initialized."

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 39]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * the first time.
  *
- * ## Example
+ * `afterContentInit` gets called whether content children have changed or not.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: this doesn't make sense. we are not detecting changes but initializing. I would remove this paragraph.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 40]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *
+ * bootstrap(App).catch(err => console.error(err));
+ * ```
  */
 export interface AfterContentInit { afterContentInit(); }
 
 /**
- * Notify a directive when the bindings of all its content children have been checked (whether
- * they have changed or not).
+ * Implement this interface to get notified when all of the content children have been checked.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: "checked and updated"

Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: "when" -> "after every time"

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 41]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  *
- * ## Example
+ * `afterContentChecked` gets called every time the content children are checked.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: not needed. remove

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 42]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
 export interface AfterContentChecked { afterContentChecked(); }
 
 /**
- * Notify a directive when the bindings of all its view children have been checked the first time
- * (whether they have changed or not).
+ * Implement this interface to get notified when all the view children have been checked for the
+ * first time.
  *
- * ## Example
+ * `afterViewInit` gets called whether view children have changed or not.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: remove

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 43]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ * bootstrap(App).catch(err => console.error(err));
+ * ```
  */
 export interface AfterContentChecked { afterContentChecked(); }
 
 /**
- * Notify a directive when the bindings of all its view children have been checked the first time
- * (whether they have changed or not).
+ * Implement this interface to get notified when all the view children have been checked for the
+ * first time.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: adjust with similar changes as afterContentInit

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 44]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  */
 export interface AfterViewInit { afterViewInit(); }
 
 /**
- * Notify a directive when the bindings of all its view children have been checked (whether they
- * have changed or not).
+ * Implement this interface to get notified when all of the view children have been checked.
  *
- * ## Example
+ * `afterViewChecked` gets called every time the view children are checked.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: remove

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 45]
```diff
@@ -30,156 +36,390 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
+ *
+ * bootstrap(App).catch(err => console.error(err));
+ * ```
  */
 export interface AfterViewInit { afterViewInit(); }
 
 /**
- * Notify a directive when the bindings of all its view children have been checked (whether they
- * have changed or not).
+ * Implement this interface to get notified when all of the view children have been checked.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: adjust with similar changes as AfterContentChecked

**Comment by IgorMinar** [comment]
One change we should make everywhere is to at least in the first sentence always mention both directives and components (even though Component is subclass of Directive) because most users won't realize this connection.

Otherwise this looks quite good!

File: modules/angular2/src/core/compiler/interfaces.ts (hunk 34 above)
Comment (vicb) [inline modules/angular2/src/core/compiler/interfaces.ts]: I've added a sentence in the API doc but I think this is an advanced topic and should be described with more details in a cookbook recipe.

File: modules/angular2/src/core/compiler/interfaces.ts (hunk 35 above)
Comment (vicb) [inline modules/angular2/src/core/compiler/interfaces.ts]: > how does doCheck interact with onChanges? does that still get called or should the "reaction" to the detected changes be handled within the doCheck method?

I'll add a note

> also does doCheck need to signal back to angular if any changes have been found?

That's the plan but it's not implemented yet (at some point it will return a boolean)

File: modules/angular2/src/core/compiler/interfaces.ts (hunk 39 above)
Comment (vicb) [inline modules/angular2/src/core/compiler/interfaces.ts]: I think it's valuable, the children might not have any data-bound properties

File: modules/angular2/src/core/compiler/interfaces.ts (hunk 41 above)
Comment (vicb) [inline modules/angular2/src/core/compiler/interfaces.ts]: I like to have a short sentence first that describe the general idea and then give more details. Do you really want those 2 previous changes ?

**Comment by vicb** [comment]
> One change we should make everywhere is to at least in the first sentence always mention both directives and components (even though Component is subclass of Directive) because most users won't realize this connection.

We should have a convention that we use across the code base. Could we handle that a part of an other PR once we have decided ?

**Comment by vicb** [comment]
@IgorMinar Thanks for the review. Most of your feeback has been implemented, I've added comments otherwise. Please let me know if that's ok.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 46]
```diff
@@ -30,156 +36,399 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
  */
 export interface AfterContentInit { afterContentInit(); }
 
 /**
- * Notify a directive when the bindings of all its content children have been checked (whether
- * they have changed or not).
+ * Implement this interface to get notified after all of the content children have been checked.
  *
- * ## Example
+ * `afterContentChecked` gets called every time the content children are checked.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: the only difference between this sentence the the sentence above is the the first one says when ("after") and the other says how often ("every time"). Otherwise the sentences are the same, so that's why I think that they should be merged into one.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 47]
```diff
@@ -30,156 +36,399 @@ export var LIFECYCLE_HOOKS_VALUES = [
...
 export interface OnDestroy { onDestroy(); }
 
 /**
- * Notify a directive when the bindings of all its content children have been checked the first
- * time (whether they have changed or not).
+ * Implement this interface to get notified when all of the content children have had their
+ * data-bound properties initialized.
  *
- * ## Example
+ * `afterContentInit` gets called whether content children have changed or not.
```
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/interfaces.ts]: the thing is that they don't "change", they get "initialized", which is what what the first sentence says. to me this looks redundant and only confuses the reader.

**Comment by IgorMinar** [comment]
replied to your comments.

wrt directive/component we can do that in a different pass.

**Comment by vicb** [comment]
PR updated, PTAL.

**Comment by IgorMinar** [comment]
I went over all of the first sentences for all of these hooks and I'm proposing the following changes (using code formatting to make visual diffing easier):

```
# OnChanges
- Implement this interface to get notified when any data-bound property of the directive changes.
+ Implement this interface to get notified when any data-bound property of your directive changes.

# OnInit
- Implement this interface to get notified when the directive or's data-bound properties have been initalized.
+ Implement this interface to execute custom initialization logic after your directive's data-bound properties have been initialized.

# DoCheck
- Implement this interface to override the default change detection algorithm for the directive.
+ Implement this interface to override the default change detection algorithm for your directive.

# OnDestroy
- Implement this interface to get notified when the directive is destroyed.
+ Implement this interface to get notified when your directive is destroyed.

# AfterContentInit
- Implement this interface to get notified when all of the content children have had their data-bound properties initialized.
+ Implement this interface to get notified when your directive's content has been fully initialized.

# AfterContentChecked
- Implement this interface to get notified each time the content children have been checked.
- Implement this interface to get notified after every check of your directive's content.

# AfterViewInit
- Implement this interface to get notified when all of the view children have had their data-bound properties initialized.
+ Implement this interface to get notified when your component's view has been fully initialized.

# AfterViewChecked
- Implement this interface to get notified each time the view children have been checked.
+ Implement this interface to get notified after every check of your component's view.
```

Sorry for all the back and forth but this api is important.

btw I'm not too happy with the "view" and "content" terminology used here as it's not consistent with the rest of the apis, but let's leave that for another discussion/pr.

**Comment by vicb** [comment]
> Sorry for all the back and forth but this api is important.

No problem I love feedback ;)
Your proposals are valuable, I'll go ahead and implement it.

WRT "view" and "content", I do agree. This is something that we should agree at global level & document.

**Comment by vicb** [comment]
@IgorMinar As much as I love feedback, I hope you'll be fine with the latest version of this PR.

_Thanks for your suggestions, all of them have been implemented_

**Comment by IgorMinar** [comment]
lgmt. go ahead and merge
